core/trino-main/src/main/java/io/trino/connector/CatalogConnector.java

```diff
@@ -33,22 +38,34 @@ public class CatalogConnector
     private final ConnectorServices systemConnector;
     private final Optional<CatalogProperties> catalogProperties;
     private final Catalog catalog;
+    private final LocalMemoryManager localMemoryManager;
 
     public CatalogConnector(
             CatalogHandle catalogHandle,
             ConnectorName connectorName,
             ConnectorServices catalogConnector,
             ConnectorServices informationSchemaConnector,
             ConnectorServices systemConnector,
+            LocalMemoryManager localMemoryManager,
             Optional<CatalogProperties> catalogProperties)
     {
         this.catalogHandle = requireNonNull(catalogHandle, "catalogHandle is null");
         this.connectorName = requireNonNull(connectorName, "connectorName is null");
         this.catalogConnector = requireNonNull(catalogConnector, "catalogConnector is null");
         this.informationSchemaConnector = requireNonNull(informationSchemaConnector, "informationSchemaConnector is null");
         this.systemConnector = requireNonNull(systemConnector, "systemConnector is null");
+        this.localMemoryManager = requireNonNull(localMemoryManager, "localMemoryManager is null");
         this.catalogProperties = requireNonNull(catalogProperties, "catalogProperties is null");
 
+        MemoryPool memoryPool = localMemoryManager.getMemoryPool();
+        long bytes = catalogConnector.getConnector().getExtraMemoryRequirement();
+        boolean success = memoryPool.tryReserveConnectorMemory(bytes);
+        if (!success) {
+            String info = format("Connector: %s, Allocated: %s, Delta: %s",
+                    connectorName, succinctBytes(memoryPool.getReservedBytes()), succinctBytes(bytes));
+            throw exceededLocalUserMemoryLimit(succinctBytes(memoryPool.getMaxBytes()), info);
```

Comment:
this should really be `exceededGlobalTotalLimit`, but with different message. Currently it's
```
format("Query exceeded distributed total memory limit of %s", maxMemory)
```
and here it's not about any particular query mem limit

Reply:
you need `exceededConnectorMemoryLimit` with new error code `EXCEEDED_CONNECTOR_MEMORY_LIMIT`

Reply:
Why global? Shouldn't it be allocated to each worker?

Reply:
Look at https://github.com/trinodb/trino/pull/22197#discussion_r1647429495. I think we need separate category